apis/objectstorage/v1alpha2/types.go

```diff
@@ -0,0 +1,336 @@
+type BucketList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Bucket `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+type BucketClaim struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec BucketClaimSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status BucketClaimStatus `json:"status,omitempty"`
+}
+
+type BucketClaimSpec struct {
+	// Name of the BucketClass
+	BucketClassName string `json:"bucketClassName,omitempty"`
+
+	// Protocols are the set of data API this bucket is required to support.
+	// The possible values for protocol are:
+	// -  S3: Indicates Amazon S3 protocol
+	// -  Azure: Indicates Microsoft Azure BlobStore protocol
+	// -  GCS: Indicates Google Cloud Storage protocol
+	Protocols []Protocol `json:"protocols"`
+
+	// Name of a bucket object that was manually
+	// created to import a bucket created outside of COSI
+	// If unspecified, then a new Bucket will be dynamically provisioned
+	// +optional
+	ExistingBucketName string `json:"existingBucketName,omitempty"`
+}
+
+type BucketClaimStatus struct {
+	// BucketReady indicates that the bucket is ready for consumpotion
+	// by workloads
+	BucketReady bool `json:"bucketReady"`
+
+	// BucketName is the name of the provisioned Bucket in response
+	// to this BucketClaim. It is generated and set by the COSI controller
+	// before making the creation request to the OSP backend.
+	// +optional
+	BucketName string `json:"bucketName,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type BucketClaimList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BucketClaim `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:storageversion
+type BucketClass struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// DriverName is the name of driver associated with this bucket
+	DriverName string `json:"driverName"`
+
+	// DeletionPolicy is used to specify how COSI should handle deletion of this
+	// bucket. There are 2 possible values:
+	//  - Retain: Indicates that the bucket should not be deleted from the OSP
+	//  - Delete: Indicates that the bucket should be deleted from the OSP
+	//        once all the workloads accessing this bucket are done
+	// +kubebuilder:default:=Retain
+	DeletionPolicy DeletionPolicy `json:"deletionPolicy"`
+
+	// Quotas defines optional quotas applied to all buckets created from this Bucket Class.
+	// +optional
+	Quotas BucketQuotas `json:"quotas,omitempty"`
```

Comment:
IMO the BucketClass should have quota based on no of buckets as well

Reply:
Is that something that can be controlled via Kubernetes RBAC?